propolis-server/src/main.rs

```diff
@@ -0,0 +1,351 @@
+                "Failed to initialize machine: {}",
+                err.to_string()
+            ))
+        })?;
+
+    instance.print();
+    instance.on_transition(Box::new(|next_state| {
+        println!("state cb: {:?}", next_state);
+    }));
+
+    // Save the newly created instance in the server's context.
+    *context =
+        Some(InstanceContext { instance, properties, serial: com1.unwrap() });
+
+    Ok(HttpResponseCreated(api::InstanceEnsureResponse {}))
+}
+
+#[endpoint {
+    method = GET,
+    path = "/instances/{instance_id}",
+}]
+async fn instance_get(
+    rqctx: Arc<RequestContext<ServerContext>>,
+    path_params: Path<api::InstancePathParams>,
+) -> Result<HttpResponseOk<api::InstanceGetResponse>, HttpError> {
+    let context = rqctx.context().context.lock().await;
+
+    let context = context.as_ref().ok_or_else(|| {
+        HttpError::for_internal_error(
+            "Server not initialized (no instance)".to_string(),
+        )
+    })?;
+
+    if path_params.into_inner().instance_id != context.properties.id {
+        return Err(HttpError::for_internal_error(
+            "UUID mismatch (path did not match struct)".to_string(),
+        ));
+    }
+    let instance_info = api::Instance {
+        properties: context.properties.clone(),
+        state: propolis_to_api_state(context.instance.current_state()),
+
+        // TODO(scaffolding)
+        disks: vec![],
+        nics: vec![],
+    };
+
+    Ok(HttpResponseOk(api::InstanceGetResponse { instance: instance_info }))
+}
+
+// TODO: Instance delete. What happens to the server? Does it shut down?
+
+#[endpoint {
+    method = PUT,
+    path = "/instances/{instance_id}/state",
+}]
+async fn instance_state_put(
+    rqctx: Arc<RequestContext<ServerContext>>,
+    path_params: Path<api::InstancePathParams>,
+    request: TypedBody<api::InstanceStateRequested>,
+) -> Result<HttpResponseUpdatedNoContent, HttpError> {
+    let context = rqctx.context().context.lock().await;
+
+    let context = context.as_ref().ok_or_else(|| {
+        HttpError::for_internal_error(
+            "Server not initialized (no instance)".to_string(),
+        )
+    })?;
+
+    if path_params.into_inner().instance_id != context.properties.id {
+        return Err(HttpError::for_internal_error(
+            "UUID mismatch (path did not match struct)".to_string(),
+        ));
+    }
+
+    let state = api_to_propolis_state(request.into_inner());
+    context.instance.set_target_state(state).map_err(|err| {
+        HttpError::for_internal_error(format!("Failed to set state: {}", err))
+    })?;
+
+    Ok(HttpResponseUpdatedNoContent {})
+}
+
+#[endpoint {
+    method = PUT,
+    path = "/instances/{instance_id}/serial",
+}]
+async fn instance_serial(
+    rqctx: Arc<RequestContext<ServerContext>>,
+    path_params: Path<api::InstancePathParams>,
+    request: TypedBody<api::InstanceSerialRequest>,
+) -> Result<HttpResponseOk<api::InstanceSerialResponse>, HttpError> {
```

Comment:
This is a short-polling mega hack. I *can* split it into two separate interfaces (where the client becomes a server) and both sides long-poll, but it feels like we should just invest that time in websockets instead.